### PR TITLE
CI/CD how-to code example

### PR DESCRIPTION
**Pull request by Blaizzy**

### CODE REVIEW

File: .github/workflows/CI_CD.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        with:
+          python-version: 3.8
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install -r CI_CD/requirements.txt
+      - name: Model Promotion(CI/CD)
+        run: python CI_CD/scripts/model_promotion.py
+      - name: Deployment Status
+        if: success()
```
Comment (jakubczakon) [inline .github/workflows/CI_CD.yml]: this looks artificial. 

perhaps we could simply drop this step and just have one step "Model Promotion (CI/CD)" ?

File: CI_CD/scripts/model_promotion.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+# Fetch project 
+project = neptune.get_project(
+    name = project_name,
+    api_token=os.getenv("NEPTUNE_API_TOKEN")
+)
+
+# In-prod
+# Filter in-prod run
+prod_runs_table_df = project.fetch_runs_table(tag='in-prod').to_pandas()
```
Comment (jakubczakon) [inline CI_CD/scripts/model_promotion.py]: I think there is always only one in-prod best model right?
So maybe if we fetch via tag 'champion' we would communicate the fetching of best model better (and make the logic simpler) ?

File: CI_CD/scripts/model_promotion.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+# Retrieve run
+stagging_run = neptune.init(
+    project=project_name,
+    api_token=os.getenv("NEPTUNE_API_TOKEN"),
+    run = stagging_run_id,
+    mode = 'read-only'
+)
+
+# Fetching and downloading data from Neptune
+parameters = prod_run['config/hyperparameters'].fetch()
```
Comment (jakubczakon) [inline CI_CD/scripts/model_promotion.py]: I think we should have it wrapped in a function called build_model from run metadata or something like that to make it obvious what is happening here. 

Cause, right now it seems that I am using prod parameters to load the stagging model and I am doing the same thing for both models (as I am getting them).

File: CI_CD/scripts/model_promotion.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+stagging_output = stagging_model(images)
+_, stagging_preds = torch.max(stagging_output, dim=1)
+
+stagging_score = (torch.sum(stagging_preds == labels)) / len(images)
+
+# Test stagging model score against production model score
+assert stagging_score >= prod_score, \
+    f'Staging model accuracy {round(stagging_score*100,2)} lower than threshold {round(prod_score*100,2)}%'
+
+print(f'Staging model with run_id = {stagging_run_id} has accuracy of {stagging_score*100}% that is greater than the current production-model was promoted to production')
```
Comment (jakubczakon) [inline CI_CD/scripts/model_promotion.py]: at this stage it wasn't promoted to production just yet so I'd change the was to is

**Review by jakubczakon — CHANGES_REQUESTED** [review]

File: CI_CD/scripts/model_promotion.py (hunk 2 above)
Comment (Blaizzy) [reply CI_CD/scripts/model_promotion.py]: You are correct, there is only one.

Since I already changed the docs, it makes sense.

**Comment by Blaizzy** [comment]
All the changes made!!!
Ready for final review.

**Comment by jakubczakon** [comment]
LGTM!

**Comment by Blaizzy** [comment]
If it looks good,
I'm merging it then! ✅

**Pull request merged**